system/lib/wasmfs/js_file_backend.cpp

```diff
@@ -0,0 +1,106 @@
+// Copyright 2021 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+// This file defines the JS file backend and JS file of the new file system.
+// Current Status: Work in Progress.
+// See https://github.com/emscripten-core/emscripten/issues/15041.
+
+#include "backend.h"
+#include "wasmfs.h"
+
+namespace wasmfs {
+
+using js_index_t = uint32_t;
+
+// This class describes a file that lives in JS Memory
+class JSFile : public DataFile {
+  // This index indicates the location of the JS File in $wasmFS$JSMemoryFiles.
+  js_index_t index;
+
+  // JSFiles will write from a Wasm Memory buffer into a JS array defined in
+  // $wasmFS$JSMemoryFiles.
+  __wasi_errno_t write(const uint8_t* buf, size_t len, off_t offset) override {
+    EM_ASM(
+      {
+        // Resize the typed array if the length of the write buffer exceeds its
+        // capacity.
+        var size = wasmFS$JSMemoryFiles[$0].length;
+        if ($2 >= size) {
+          var oldContents = wasmFS$JSMemoryFiles[$0];
+          wasmFS$JSMemoryFiles[$0] = new Uint8Array($2);
+          wasmFS$JSMemoryFiles[$0].set(oldContents.subarray(0, $2));
+        }
+        wasmFS$JSMemoryFiles[$0].set(HEAPU8.subarray($1, $1 + $2), $3);
+      },
+      index,
+      buf,
+      len,
+      (size_t)offset);
```

Comment:
I think now might be a good time to follow the suggestions from a few weeks ago and use something other than EM_ASM. EM_ASM isn't that great in general for code size, but more specifically here, once you start to use things from JS like `wasmFS$JSMemoryFiles` then it's more readable to put the code next to those definitions, that is, to put it all on the JS side so it's not split over separate files.

Another benefit is that you don't get $0, $1 etc. names for parameters - you can define actually readable names. That is also possible with EM_JS, which is another option here. Overall, EM_JS would keep things in this file, which is good if there is tight integration with the calling C++, but that doesn't seem to be the case here.

Reply:
Done.